src/nodes/NodeConnectionManager.ts

```diff
@@ -1378,106 +1400,129 @@ class NodeConnectionManager {
+    if (!this.rateLimiter.consume(sourceNodeIdString)) {
+      throw new nodesErrors.ErrorNodeConnectionManagerRequestRateExceeded();
+    }
+    // Generating relay signature, data is just `<sourceNodeId><targetNodeId><Address><requestSignature>` concatenated
+    const data = Buffer.concat([
+      sourceNodeId,
+      targetNodeId,
+      Buffer.from(JSON.stringify(address), 'utf-8'),
+      Buffer.from(requestSignature, 'base64url'),
+    ]);
+    const relaySignature = keysUtils.signWithPrivateKey(
+      this.keyRing.keyPair,
+      data,
     );
-    // Send message and ignore any error
-    await this.withConnF(
-      relayNodeId,
-      async (connection) => {
-        const client = connection.getClient();
-        await client.methods.nodesHolePunchMessageSend(
-          {
-            srcIdEncoded: srcNode,
-            dstIdEncoded: tgtNode,
-            address,
-          },
-          ctx,
-        );
-      },
-      ctx,
-    ).catch(() => {});
+    const connProm = this.withConnF(targetNodeId, async (conn) => {
+      const client = conn.getClient();
+      await client.methods.nodesConnectionSignalFinal({
+        sourceNodeIdEncoded: nodesUtils.encodeNodeId(sourceNodeId),
+        targetNodeIdEncoded: nodesUtils.encodeNodeId(targetNodeId),
+        address,
+        requestSignature: requestSignature,
+        relaySignature: relaySignature.toString('base64url'),
+      });
+    }).finally(() => {
+      this.activeSignalSet.delete(connProm);
+    });
+    this.activeSignalSet.add(connProm);
   }
 
   /**
-   * Forwards a received hole punch message on to the target.
-   * If not known, the node ID -> address mapping is attempted to be discovered
-   * through Kademlia (note, however, this is currently only called by a 'broker'
-   * node).
-   * @param message the original relay message (assumed to be created in
-   * nodeConnection.start())
-   * @param sourceAddress
+   * This till ask a signalling node to signal a target node to hole punch back to this node.
+   * @param targetNodeId - NodeId of the node that needs to signal back.
+   * @param signallingNodeId - NodeId of the signalling node.
    * @param ctx
    */
-  public relaySignalingMessage(
-    message: HolePunchRelayMessage,
-    sourceAddress: NodeAddress,
-    ctx?: Partial<ContextTimed>,
+  public holePunchSignalRequest(
+    targetNodeId: NodeId,
+    signallingNodeId: NodeId,
+    ctx?: Partial<ContextTimedInput>,
   ): PromiseCancellable<void>;
-  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  @ready(new nodesErrors.ErrorNodeManagerNotRunning())
   @timedCancellable(
     true,
     (nodeConnectionManager: NodeConnectionManager) =>
       nodeConnectionManager.connectionConnectTimeoutTime,
   )
-  public async relaySignalingMessage(
-    message: HolePunchRelayMessage,
-    sourceAddress: NodeAddress,
+  public async holePunchSignalRequest(
+    targetNodeId: NodeId,
+    signallingNodeId: NodeId,
```

Comment:
I think the naming of this should correlate to "initial" and "final". Otherwise it is hard to understand and correlate them to request/forward/reverse.

Reply:
Changed the name and updated the docblock.